File: Scala/test-1/SpatialLogic.scala
object SpatialLogic {
  case class Point(pointX: Double, pointY: Double)

  case class Polygon(
      polygonVertices: Vector[Point],
      polygonNumVertices: Int,
      polygonIsConvex: Boolean,
      polygonZoneName: String
  )

  case class SpatialLayer(
      layerPolygons: Vector[Polygon],
      layerPolygonCount: Int,
      layerName: String
  )

  def calculateCleanMinkowskiSum(
      poly1: Option[Polygon],
      poly2: Option[Polygon]
  ): Option[Polygon] = sys.error("SpatialLogic.calculateCleanMinkowskiSum not implemented")

  def checkDynamicCollision(
      obstacle: Option[Polygon],
      robot: Option[Polygon],
      velocity: Point
  ): Boolean = sys.error("SpatialLogic.checkDynamicCollision not implemented")
}

File: Haskell/test-1/SpatialLogic.hs
module SpatialLogic
  ( Point (..)
  , Polygon (..)
  , SpatialLayer (..)
  , calculateCleanMinkowskiSum
  , checkDynamicCollision
  ) where

data Point = Point
  { pointX :: Double
  , pointY :: Double
  } deriving (Eq, Show)

data Polygon = Polygon
  { polygonVertices :: [Point]
  , polygonNumVertices :: Int
  , polygonIsConvex :: Bool
  , polygonZoneName :: String
  } deriving (Eq, Show)

data SpatialLayer = SpatialLayer
  { layerPolygons :: [Polygon]
  , layerPolygonCount :: Int
  , layerName :: String
  } deriving (Eq, Show)

calculateCleanMinkowskiSum :: Maybe Polygon -> Maybe Polygon -> Maybe Polygon
calculateCleanMinkowskiSum _ _ = error "SpatialLogic.calculateCleanMinkowskiSum not implemented"

checkDynamicCollision :: Maybe Polygon -> Maybe Polygon -> Point -> Bool
checkDynamicCollision _ _ _ = error "SpatialLogic.checkDynamicCollision not implemented"
